tests/suite/upgrade_test.go

```diff
@@ -0,0 +1,246 @@
+		resultsFile.Close()
+	})
+
+	It("upgrades NGF with zero downtime", func() {
+		cfg := framework.InstallationConfig{
+			ReleaseName:          releaseName,
+			Namespace:            ngfNamespace,
+			ChartPath:            localChartPath,
+			NgfImageRepository:   *ngfImageRepository,
+			NginxImageRepository: *nginxImageRepository,
+			ImageTag:             *imageTag,
+			ImagePullPolicy:      *imagePullPolicy,
+			ServiceType:          *serviceType,
+			IsGKEInternalLB:      *isGKEInternalLB,
+		}
+
+		type metricsResults struct {
+			metrics  *vegeta.Metrics
+			testName string
+			scheme   string
+		}
+		metricsCh := make(chan *metricsResults)
+
+		type testCfg struct {
+			desc   string
+			port   string
+			target framework.Target
+		}
+
+		tests := []testCfg{
+			{
+				desc: "Send http /coffee traffic",
+				port: "80",
+				target: framework.Target{
+					Method: "GET",
+					URL:    "http://cafe.example.com/coffee",
+				},
+			},
+			{
+				desc: "Send https /tea traffic",
+				port: "443",
+				target: framework.Target{
+					Method: "GET",
+					URL:    "https://cafe.example.com/tea",
+				},
+			},
+		}
+
+		for _, test := range tests {
+			go func(cfg testCfg) {
+				defer GinkgoRecover()
+
+				results, metrics := framework.RunLoadTest(
+					[]framework.Target{cfg.target},
+					100,
+					60*time.Second,
+					cfg.desc,
+					fmt.Sprintf("%s:%s", address, cfg.port),
+					"cafe.example.com",
+				)
+
+				scheme := strings.Split(cfg.target.URL, "://")[0]
+				metricsRes := metricsResults{
+					metrics:  &metrics,
+					testName: fmt.Sprintf("\n## Test: %s\n\n```text\n", cfg.desc),
+					scheme:   scheme,
+				}
+
+				buf := new(bytes.Buffer)
+				encoder := vegeta.NewCSVEncoder(buf)
+				for _, res := range results {
+					res := res
+					Expect(encoder.Encode(&res)).To(Succeed())
+				}
+
+				csvName := fmt.Sprintf("%s.csv", scheme)
+				filename := filepath.Join(resultsDir, csvName)
+				csvFile, err := framework.CreateResultsFile(filename)
+				Expect(err).ToNot(HaveOccurred())
+
+				_, err = fmt.Fprint(csvFile, buf.String())
+				Expect(err).ToNot(HaveOccurred())
+				csvFile.Close()
+
+				output, err := framework.GeneratePNG(resultsDir, csvName, fmt.Sprintf("%s.png", scheme))
+				Expect(err).ToNot(HaveOccurred(), string(output))
+
+				metricsCh <- &metricsRes
+			}(test)
+		}
+
+		// allow traffic flow to start
+		time.Sleep(2 * time.Second)
+
+		output, err := framework.UpgradeNGF(cfg, "--values", valuesFile)
+		Expect(err).ToNot(HaveOccurred(), string(output))
+
+		Expect(resourceManager.ApplyFromFiles([]string{"ngf-upgrade/gateway-updated.yaml"}, ns.Name)).To(Succeed())
+
+		podNames, err := framework.GetNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+		Expect(err).ToNot(HaveOccurred())
+
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+		defer cancel()
+
+		// ensure that the leader election lease has been updated to the new pods
```

Comment:
Why not poll here? Leader election change may not happen immediately.

Reply:
I originally did poll, but noticed that the leader election always switched over before the upgrade call returned, which is why I removed it.

Reply:
How many times did you run the tests? Is there any harm in adding a poll to cover the possibility of leader election change taking longer?

Reply:
I've run it 20+ times at this point. No harm in it, but it has yet to fail.

Reply:
lol that is a lot of times. We can leave it as is.

Reply:
Now that I say this, it'll fail first time in the pipeline :)